File: packages/typescript/src/modules/customers.ts
/**
 * CustomersModule: getOrCreate and update customer IDs for tokenisation.
 */

import { mapReturnCodeToError, CustomerException } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { BASE_URLS } from '../config.js';
import type {
  GetCustomerIdRequest,
  GetCustomerIdResponse,
  PaymentInfoItem,
} from '../types/api.js';
import type { CustomerInfo, EcollectCustomer } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';

function info(code: number, desc: string, value: string | undefined): PaymentInfoItem | null {
  if (!value) return null;
  return { AttributeCode: code, AttributeDesc: desc, AttributeValue: value };
}

export class CustomersModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
  }

  /**
   * Get or create a CustomerId for a customer.
   * Returns existing CustomerId if the customer already exists.
   */
  async getOrCreateCustomerId(customerInfo: CustomerInfo): Promise<EcollectCustomer> {
    const url = `${BASE_URLS[this.config.environment]}/getCustomerId`;
    const sessionToken = await this.session.getActive();

    const customerInfoArray: PaymentInfoItem[] = [
      { AttributeCode: 6, AttributeDesc: 'Usermail', AttributeValue: customerInfo.email },
      { AttributeCode: 19, AttributeDesc: 'CardHolderId', AttributeValue: customerInfo.documentNumber },
      { AttributeCode: 18, AttributeDesc: 'CardHolderIdType', AttributeValue: customerInfo.documentType },
      { AttributeCode: 17, AttributeDesc: 'CardHolderName', AttributeValue: customerInfo.fullName },
      { AttributeCode: 7, AttributeDesc: 'MobileCountryCode', AttributeValue: customerInfo.mobileCountryCode },
      { AttributeCode: 8, AttributeDesc: 'MobileNumber', AttributeValue: customerInfo.mobileNumber },
    ].filter((x) => x.AttributeValue) as PaymentInfoItem[];

    const body: GetCustomerIdRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
      CustomerInfoArray: customerInfoArray,
    };

    const res = await this.http.post<GetCustomerIdRequest, GetCustomerIdResponse>(url, body);

    if (res.ReturnCode === 'FAIL_APIEXPIREDSESSION') {
      this.session.invalidate();
      body.SessionToken = await this.session.getActive();
      const retried = await this.http.post<GetCustomerIdRequest, GetCustomerIdResponse>(url, body);
      if (retried.ReturnCode !== 'SUCCESS') {
        throw mapReturnCodeToError(retried.ReturnCode);
      }
      return this._parseCustomer(retried, customerInfo);
    }

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return this._parseCustomer(res, customerInfo);
  }

  /**
   * Update customer information by providing a CustomerId.
   */
  async updateCustomerInfo(
    customerId: string,
    updatedInfo: Partial<CustomerInfo>,
  ): Promise<EcollectCustomer> {
    const url = `${BASE_URLS[this.config.environment]}/getCustomerId`;
    const sessionToken = await this.session.getActive();

    const items: Array<PaymentInfoItem | null> = [
      { AttributeCode: 100, AttributeDesc: 'CustomerId', AttributeValue: customerId },
      info(6, 'Usermail', updatedInfo.email),
      info(17, 'CardHolderName', updatedInfo.fullName),
      info(18, 'CardHolderIdType', updatedInfo.documentType),
      info(7, 'MobileCountryCode', updatedInfo.mobileCountryCode),
      info(8, 'MobileNumber', updatedInfo.mobileNumber),
    ];

    const customerInfoArray = items.filter((x): x is PaymentInfoItem => x !== null);

    const body: GetCustomerIdRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
      CustomerInfoArray: customerInfoArray,
    };

    const res = await this.http.post<GetCustomerIdRequest, GetCustomerIdResponse>(url, body);

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return this._parseCustomer(res, { ...updatedInfo, documentNumber: '' } as CustomerInfo);
  }

  private _parseCustomer(
    res: GetCustomerIdResponse,
    original: Partial<CustomerInfo>,
  ): EcollectCustomer {
    const arr = res.CustomerInfoArray ?? [];
    const findAttr = (code: number) => arr.find((i) => i.AttributeCode === code)?.AttributeValue;

    const customerId = findAttr(100);
    if (!customerId) {
      throw new CustomerException('getCustomerId response did not return a CustomerId');
    }

    return {
      customerId,
      email: original.email,
      fullName: original.fullName,
      documentType: original.documentType,
      documentNumber: original.documentNumber,
      mobileCountryCode: original.mobileCountryCode,
      mobileNumber: original.mobileNumber,
    };
  }
}

File: packages/python/ecollect/modules/customers.py
"""Customer management module (getCustomerId)."""
import logging
from typing import Any, Dict, Optional

from ecollect.config import EcollectConfig
from ecollect.exceptions import raise_for_return_code
from ecollect.utils.http import AsyncHttpClient, run_sync

logger = logging.getLogger(__name__)


def _info_item(code: int, desc: str, value: str) -> Dict[str, Any]:
    return {"AttributeCode": code, "AttributeDesc": desc, "AttributeValue": str(value)}


class CustomersModule:
    """Handles customer identity management for tokenisation."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session

    async def _post(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        url = self._config.endpoint_url(endpoint)
        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        if return_code == "FAIL_APIEXPIREDSESSION":
            self._session.invalidate()
            payload["SessionToken"] = await self._session.get_active()
            data = await self._http.post_with_retry(url, payload)
        raise_for_return_code(data.get("ReturnCode", ""), str(data))
        return data

    # ------------------------------------------------------------------
    # Async API
    # ------------------------------------------------------------------

    async def get_or_create_customer_id(
        self,
        email: str,
        card_holder_id: str,
        card_holder_name: str,
        card_holder_id_type: str,
        mobile_country_code: str,
        mobile_number: str,
    ) -> str:
        """Get or create a persistent CustomerId for a customer.

        Returns the CustomerId string from ecollect.
        """
        session_token = await self._session.get_active()
        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "CustomerInfoArray": [
                _info_item(6, "Usermail", email),
                _info_item(19, "CardHolderId", card_holder_id),
                _info_item(17, "CardHolderName", card_holder_name),
                _info_item(18, "CardHolderIdType", card_holder_id_type),
                _info_item(7, "MobileCountryCode", mobile_country_code),
                _info_item(8, "MobileNumber", mobile_number),
            ],
        }
        data = await self._post("getCustomerId", payload)
        # CustomerId is returned in CustomerInfoArray
        for item in data.get("CustomerInfoArray", []):
            if item.get("AttributeDesc", "").lower() == "customerid":
                return item["AttributeValue"]
        # Fallback: look for AttributeCode 36 (CustomerId)
        for item in data.get("CustomerInfoArray", []):
            if item.get("AttributeCode") == 36:
                return item["AttributeValue"]
        return data.get("CustomerId", "")

    async def update_customer_info(
        self,
        customer_id: str,
        email: Optional[str] = None,
        card_holder_name: Optional[str] = None,
        card_holder_id_type: Optional[str] = None,
        mobile_country_code: Optional[str] = None,
        mobile_number: Optional[str] = None,
    ) -> str:
        """Update an existing customer's information.

        Returns the CustomerId (unchanged).
        """
        session_token = await self._session.get_active()
        info_array = [_info_item(36, "CustomerId", customer_id)]
        if email:
            info_array.append(_info_item(6, "Usermail", email))
        if card_holder_name:
            info_array.append(_info_item(17, "CardHolderName", card_holder_name))
        if card_holder_id_type:
            info_array.append(_info_item(18, "CardHolderIdType", card_holder_id_type))
        if mobile_country_code:
            info_array.append(_info_item(7, "MobileCountryCode", mobile_country_code))
        if mobile_number:
            info_array.append(_info_item(8, "MobileNumber", mobile_number))

        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "CustomerInfoArray": info_array,
        }
        data = await self._post("getCustomerId", payload)
        return customer_id

    # ------------------------------------------------------------------
    # Sync wrappers
    # ------------------------------------------------------------------

    def get_or_create_customer_id_sync(self, *args, **kwargs) -> str:
        return run_sync(self.get_or_create_customer_id(*args, **kwargs))

    def update_customer_info_sync(self, *args, **kwargs) -> str:
        return run_sync(self.update_customer_info(*args, **kwargs))
